File: Ant/chapter3/python/stock_charts.py
def ns(f):
    return next(f).strip()


with open("../testset/stock_charts/test1.txt", 'r') as f:
    N, K = map(int, ns(f).split())
    P = [list(map(int, ns(f).split())) for _ in range(N)]

V = N * 2
G = [[] for _ in range(V)]


def add_edge(u, v):
    global G
    G[u].append(v)
    G[v].append(u)


def dfs(u, used, match):
    used[u] = True
    for v in G[u]:
        _u = match[v]
        if _u < 0 or (not used[_u] and dfs(_u, used, match)):
            match[v] = u
            match[u] = v
            return True
    return False


def bipartite_matching():
    res = 0
    match = [-1] * V
    for u in range(N):
        if match[u] < 0:
            used = [False] * V
            if dfs(u, used, match):
                res += 1
    return res


for i in range(N):
    for j in range(N):
        if i == j:
            continue
        f = True
        for k in range(K):
            if P[j][k] >= P[i][k]:
                f = False
                break
        if f:
            add_edge(i, N + j)
print(N - bipartite_matching())

File: Ant/chapter3/cpp/stock_charts.cpp
#include <fstream>
#include <memory.h>
#include <stdio.h>
#include <vector>

#define MAX_N 100
#define MAX_V MAX_N * 2
#define MAX_K 25

using namespace std;

int N, K, V;
int P[MAX_N][MAX_K];
vector<int> G[MAX_V];
int match[MAX_V];
bool used[MAX_V];

void add_edge(int u, int v) {
  G[u].push_back(v);
  G[v].push_back(u);
}

bool dfs(int u) {
  used[u] = true;
  for (auto v : G[u]) {
    int _u = match[v];
    if (_u < 0 || (!used[_u] && dfs(_u))) {
      match[v] = u;
      match[u] = v;
      return true;
    }
  }
  return false;
}

int bipartite_matching() {
  int res = 0;
  memset(match, -1, sizeof(match));
  for (int u = 0; u < N; ++u) {
    if (match[u] < 0) {
      memset(used, false, sizeof(used));
      if (dfs(u)) {
        ++res;
      }
    }
  }
  return res;
}

void solve() {
  V = N * 2;
  for (int i = 0; i < N; ++i) {
    for (int j = 0; j < N; ++j) {
      if (i == j) {
        continue;
      }
      bool f = true;
      for (int k = 0; k < K; ++k) {
        if (P[j][k] >= P[i][k]) {
          f = false;
          break;
        }
      }
      if (f) {
        add_edge(i, N + j);
      }
    }
  }
  printf("%d\n", N - bipartite_matching());
}

int main() {
  ifstream ifs("../testset/stock_charts/test1.txt");
  ifs >> N >> K;
  for (int i = 0; i < N; ++i) {
    for (int j = 0; j < K; ++j) {
      ifs >> P[i][j];
    }
  }
  solve();
}